content/blog/elm-warrior-walkthrough/index.md

```diff
@@ -0,0 +1,818 @@
+        |> Maybe.withDefault Warrior.Wait
+```
+
+Unfortunately this is the result...
+
+![Elm warrior level three oopsie](./level-three-oopsie.gif)
+
+The problem with our approach is that we're using a stateless algorithm and
+we're not checking if we already visited the same tile in the past. Now
+let's take another look at the `History` module. We'll find this function:
+
+```elm
+{-| Returns a list of the state of a warrior and the map,
+going back to the first turn. The first element in the list
+will represent the state of the warrior and the map at the
+beginning of that warrior's last turn.
+-}
+previousStates : Warrior -> History -> List ( Warrior, Map )
+```
+
+What we get is a list of all previous states of our warrior, along with the
+state of the map at the time. So we can use the `position` function from
+the `Warrior` module to detect if we ever visited that tile before.
+
+```elm
+takeTurn : Warrior -> Map -> History -> Warrior.Action
+takeTurn warrior map history =
+    let
+        alreadyVisited coords =
+            History.previousStates warrior history
+                |> List.any
+                    (\( w, _ ) ->
+                        Warrior.position w == coords
+                    )
+
+        checkDirection dir =
+            case Map.look dir warrior map of
+                ( newCoords, Empty ) :: _ ->
+                    if alreadyVisited newCoords then
+                        Nothing
+
+                    else
+                        Just (Warrior.Move dir)
+
+                ( _, Exit ) :: _ ->
+                    Just (Warrior.Move dir)
+
+                _ ->
+                    Nothing
+    in
+    [ Right, Down, Left, Up ]
+        |> List.foldl
+            (\dir acc ->
+                case acc of
+                    Just v ->
+                        acc
+
+                    Nothing ->
+                        checkDirection dir
+            )
+            Nothing
+        |> Maybe.withDefault Warrior.Wait
+```
+
+Tada! 🎉
+
+## Level IV
+
+So [here](https://ellie-app.com/9mqpJDnTq8ca1) is where we are now.
+
+![Elm warrior level four](./level-four.png)
+
+Our warrior is now stuck on the top right corner. Why is that? Well, we
+just added some logic that makes the warrior skip tiles which have been
+visited in the past. In this case, we went down a path which turned out to
+be a dead end, so we **need to backtrack**.
+
+So instead of completely avoiding to step on a tile that has already been
+visited, we can think which are the possibilities when we want to explore
+the world in a certain direction. We can find:
+
+- the exit tile
+- an empty unvisited tile
+- an empty already visited tile
+- a wall tile
+
+So what we can do is to assign a value to each of these possibilities, so
+that we can choose the best available move every time. The great thing
+about Elm is how easy it is to create types for everything, so let's do
+that.
+
+```elm
+type Exploration
+    = Freedom
+    | CanMove Int
```

Comment:
I think there's a big gap between "I want to avoid visited tiles" and "I'll keep a counter of the times we visited each and always prefer the one with the lowest count".

Like, I felt that a boolean would be the obvious next step at this point and when I read this I stopped to check if I was missing something. Adding a comment about why you went straight here would be really nice IMO.

Reply:
Yup, added a small intermediate representation so no links are missing :)